MSI Authentication

sebastiankasprzak opened the pull request:
Implementing #57 

### Overview
MSI overview: https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview

This PR introduces option for MSI authentication as an alternative to SPN credentials. For VM's deployed in Azure, once MSI is enabled, then it is possible to authenticate with Azure mgmt API and KeyVault using MSI, therefore it is no logner required to generate and store Service Principal Credentials.

This may be considered as a preferred options for users interested in using this Splunk AddOn but are concerned about security implications of having to use SPN credentials.

Changes are fully backwards compatible with SPN method. If `SPNApplicationId` and `SPNApplicationKey` are not specified in inputs.conf file, the app will try to authenticate using MSI.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/AzureMonitorAddonForSplunk?pullRequest=61) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/AzureMonitorAddonForSplunk?pullRequest=61) before we can accept your contribution.<br/><br/>:x: sebastiankasprzak [sign now](https://cla.opensource.microsoft.com/Microsoft/AzureMonitorAddonForSplunk?pullRequest=61)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/AzureMonitorAddonForSplunk?pullRequest=61) it.</sub>

sebastiankasprzak commented:
@rickrain @sebastus would you be able to take a look at my PR and let me know what you think?

sebastus commented:
@sebastiankasprzak I have looked it over. It's a big PR, so will take some time to review thoroughly. From what I've seen, it looks really great! One thing that jumps out at me is wiki updates. We'll need some. Also, I'm doing another chunk of work (AAD integration) that's going to supersede v1.2.9, so yours will be 1.3.0 most likely.

FYI, neither Rick nor I are on this full time. It's best effort - we have primary job functions that take precedence. 

@Rickrain there are setup script updates as well as program updates.

Thanks a lot for your contribution and interest!

sebastiankasprzak commented:
Thank you for quick response @sebastus. 

I appreciate the other priorities and the best effort nature of the project, I will monitor the PR and also I will take a look at updating the wiki in the meantime. 

Let me know if I can help in any other way.

sebastus commented:
In azure_monitor_metrics_main.py, can we stick lines 208-218 inside the if statement at line 220? I'll take a look at your wiki changes.

sebastiankasprzak commented:
@sebastus I moved the lines over, tested and all seems good on my end after the change

sebastus commented:
Found this in the log while testing with "normal" creds (not MSI).

08-16-2018 13:44:45.315 +0000 INFO  ExecProcessor - message from "/opt/splunk/etc/apps/TA-Azure_Monitor/bin/azure_diagnostic_logs.sh" Modular input Retrieving SAS credentials for Eventhub from KeyVault undefined

Found:

Logger.info("Retrieving SAS credentials for Eventhub from KeyVault")

in subs.js, line 64. Logger.info requires the name of the TA in the first parameter. Check other examples of the use of Logger.error and Logger.debug. 

Can we change the Logger.info to Logger.debug? 

While I was poking around, I found Logger.info at line 100 of checkpoint.js. Would you mind switching that to error?

sebastiankasprzak commented:
@sebastus I removed the logging module from subs.js as it was not even that useful for the single log line in that file, corrected checkpoint file log level as well

sebastus commented:
I found that if the tenant Id, app id and app key are undefined, the log gets filled up with what looks like error messages. Let's add this to line 76 of azure_monitor_logs.js:

if (_.isUndefined(singleInput.SPNApplicationId) && _.isUndefined(singleInput.SPNApplicationKey)) {
        done(null, singleInput);
    } else if (singleInput….

Also - if I run the addon without setting up MSI on the VM, I get this message:

08-20-2018 11:53:47.830 +0000 ERROR ExecProcessor - message from "/opt/splunk/etc/apps/TA-Azure_Monitor/bin/azure_diagnostic_logs.sh" (node:19566) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: connect ECONNREFUSED 127.0.0.1:50342

I didn't track it down, but it would be better if the reject is handled and if you can, put in a message that gives a better hint as to what's going on.

Thanks again for all the hard work!

sebastiankasprzak commented:
apologies for delay on this, had few busy days at work. I should be back on this next week, in meantime let me know if you find anything else worth improving.

sebastus commented:
My plan is to run through various "screwed up" scenarios and add the logs necessary to help a new user figure it out for themselves. Part of this is removing logs that make no sense under the circumstances. I think the message we're currently up against is a blocker for further testing of that type.

sebastiankasprzak commented:
Apologies on delay with this. I have added additional block to catch and log issues with obtaining MSI authentication token.

I'm not seeing my logs filling up with error messages under normal circumstances though, could you provide an example you are seeing at your end @sebastus ?

sebastus commented:
re: filling up the log:
The extraneous messages are there with debug log level on ExecProcessor. You'll see these extra messages that should not be there because the spn creds aren't present:

password was deleted
creating new password
password was deleted
creating new password
New password was created
New password was created
program path and name is: /opt/splunk/bin/splunk
stdout: azure_diagnostic_logs: Error updating inputs.conf: SPNTenantID
stderr:

re: new version number
package.json also has version number in it that needs to be updated.

re: msi auth
I stepped through your instructions on setting up the MSI for the VM. I verified that MSI is working on the VM by using az cli. With that tool, I can authenticate with msi on the VM and then get the secret from the key vault. However, I still get the error message:

Modular input getMSIToken Failed to obtain MSI token:: Error: connect ECONNREFUSED 127.0.0.1:50342

Note that ms-rest-azure is using http://localhost:50342 to access msi. I see this method is going to be deprecated (https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview#how-a-system-assigned-identity-works-with-an-azure-vm). What did I miss?

sebastiankasprzak commented:
@sebastus got you now on the additional debug messages, fix implemented as per your recommendation.

package.json file updated as well

so I have been using ARM template to enable the MSI extension as part of my internal pipeline. I just realised that I'm having similar problems when using CLI command to enable MSI extension on VM, and I've noticed that results of CLI commands are inconsistent with what I'm seeing in Azure Portal. Can you check that when in Azure portal you go to your VM and then `Identity` that the status for system assigned identity is set to "on"? If not, change it to "on", save, and let me know if that resolved the issue on your end as well.

I can reach out to Azure Support to clarify on plans to update ms-rest-azure package to use the new endpoint for authentication (unless you have someone to ask on your end?)

sebastus commented:
@sebastiankasprzak The identity is "on". Has been the whole time. I don't have a specific person I could reach out to, but I can work on that. I don't think it's deprecated yet. Have you gotten it to work at all with CLI / Portal? Or just when using ARM template?

sebastiankasprzak commented:
yes, for me it works fine when I use portal or ARM to enable it, but did not manage to get it to work using only CLI. I will troubleshoot it further tomorrow though.

sebastiankasprzak commented:
@sebastus thanks for raising this. I think I got to the bottom of it:

`az vm identity assign` CLI command enables the MSI extension, however it does not deploy msi-extension agent on the VM, hence nothing is listening on port 50342.

When using ARM template or portal, when MSI gets enabled, the extension is also deployed to the VM automatically.

There is a 2nd API call required to deploy the extension:
`az vm extension set -n ManagedIdentityExtensionForLinux --publisher Microsoft.ManagedIdentity --vm-name {vm name} --resource-group {rg name}`

I'm going to add this API call to the setup script just now. 

Ideally once the `ms-rest-azure` package is updated to use metadata endpoint then the endpoint deployment will no longer be needed, alternatively we could swap the use of package in favour of raw http call, but I would rather stick to the nodejs package if possible.

sebastiankasprzak commented:
@sebastus I've now added extra cli cmd's to azure_setup bash script

sebastus commented:
Looks like we need a final build of the package?

sebastiankasprzak commented:
nice one!

package file updated.

How do you want to handle the wiki changes?
